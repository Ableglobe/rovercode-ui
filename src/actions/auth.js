// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = isValidAuth => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});
