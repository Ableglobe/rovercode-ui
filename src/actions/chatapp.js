import Axios from "axios";

export const TOGGLE_FORMS = 'TOGGLE_FORMS';
export const SET_SESSION_ID = 'SET_SESSION_ID';
export const SET_CLIENT_ID = 'SET_CLIENT_ID';
export const SET_IS_SUPPORT_PROVIDER = 'SET_IS_SUPPORT_PROVIDER';


export const toggleForms = () => ({
  type: TOGGLE_FORMS,
});

export const setClientID = id => ({
  type: SET_CLIENT_ID,
  payload: id,
});

export const setSessionID = id => ({
  type: SET_SESSION_ID,
  payload: id,
});

export const setIsSupportProvider = () => ({
  type: SET_IS_SUPPORT_PROVIDER,
});

