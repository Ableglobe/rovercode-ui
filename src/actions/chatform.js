export const CATEGORY_SELECT_CHANGE = 'CATEGORY_SELECT_CHANGE';
export const BODY_INPUT_CHANGE = 'BODY_INPUT_CHANGE';
export const EXPERIENCE_SELECT_CHANGE = 'EXPERIENCE_SELECT_CHANGE'
export const SUBJECT_INPUT_CHANGE = 'SUBJECT_INPUT_CHANGE';

// CATEGORY STATES
export const FIRST_TIME = 0;
export const PRETTY_NEW = 1;
export const FOR_A_WHILE = 2;
export const EXPERT = 3;

//EXPERIENCE STATES
export const DEBUG = "DEBUG";
export const WHAT_NEXT = "WHAT_NEXT";
export const FUN_IDEAS = "FUN_IDEAS";


export const categorySelectChange = (value) =>({
    type: CATEGORY_SELECT_CHANGE,
    payload: value
})

export const subjectInputChange = (value) =>({
    type: SUBJECT_INPUT_CHANGE,
    payload: value
})

export const bodyInputChange = (value) =>({
    type: BODY_INPUT_CHANGE,
    payload: value
})

export const experienceSelectChange = (value) =>({
    type: EXPERIENCE_SELECT_CHANGE,
    payload: value
})

