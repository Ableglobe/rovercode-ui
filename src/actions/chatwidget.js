export const TOGGLE_IN_PROGRESS_STATE = 'TOGGLE_IN_PROGRESS_STATE';

export const toggleInProgressState = () =>({
    type: TOGGLE_IN_PROGRESS_STATE,
  })