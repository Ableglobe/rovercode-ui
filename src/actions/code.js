// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_JSCODE = 'UPDATE_JSCODE';
export const CHANGE_EXECUTION_STATE = 'CHANGE_EXECUTION_STATE';

// Execution States
export const EXECUTION_RUN = 1;
export const EXECUTION_STEP = 2;
export const EXECUTION_STOP = 3;
export const EXECUTION_RESET = 4;

// action creators
export const updateJsCode = jsCode => ({
  type: UPDATE_JSCODE,
  payload: jsCode,
});

export const changeExecutionState = state => ({
  type: CHANGE_EXECUTION_STATE,
  payload: state,
});
