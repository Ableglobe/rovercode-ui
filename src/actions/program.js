// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const FETCH_USER_PROGRAMS = 'FETCH_USER_PROGRAMS';
export const FETCH_USER_PROGRAMS_PENDING = `${FETCH_USER_PROGRAMS}_PENDING`;
export const FETCH_USER_PROGRAMS_FULFILLED = `${FETCH_USER_PROGRAMS}_FULFILLED`;
export const FETCH_USER_PROGRAMS_REJECTED = `${FETCH_USER_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const fetchPrograms = (xhrOptions) => {
  let type = FETCH_PROGRAMS;

  if (xhrOptions && xhrOptions.params && xhrOptions.params.user) {
    type = FETCH_USER_PROGRAMS;
  }

  return ({
    type,
    meta: {
      debounce: {
        time: 500,
      },
    },
    payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
      .then(({ data }) => (
        data
      )),
  });
};

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
