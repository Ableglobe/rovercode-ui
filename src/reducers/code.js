import {
  CHANGE_EXECUTION_STATE,
  UPDATE_JSCODE,
  UPDATE_XMLCODE,
  CHANGE_NAME,
  CHANGE_NAME_FULFILLED,
  CHANGE_NAME_REJECTED,
  CHANGE_ID,
  FETCH_PROGRAM,
  FETCH_PROGRAM_FULFILLED,
  FETCH_PROGRAM_REJECTED,
  SAVE_PROGRAM,
  SAVE_PROGRAM_FULFILLED,
  SAVE_PROGRAM_REJECTED,
  CREATE_PROGRAM,
  CREATE_PROGRAM_FULFILLED,
  CREATE_PROGRAM_REJECTED,
} from '../actions/code';

export default function code(
  state = {
    jsCode: null,
    xmlCode: null,
    execution: null,
    name: null,
    id: null,
    isFetching: false,
    isSaving: false,
    isCreating: false,
    isChangingName: false,
    error: null,
  },
  action,
) {
  switch (action.type) {
    case UPDATE_JSCODE:
      return {
        ...state,
        jsCode: action.payload,
      };
    case UPDATE_XMLCODE:
      return {
        ...state,
        xmlCode: action.payload,
      };
    case CHANGE_EXECUTION_STATE:
      return {
        ...state,
        execution: action.payload,
      };
    case CHANGE_NAME:
      return {
        ...state,
        isChangingName: true,
      };
    case CHANGE_NAME_FULFILLED:
      return {
        ...state,
        isChangingName: false,
        name: action.payload.name,
      };
    case CHANGE_NAME_REJECTED:
      return {
        ...state,
        isChangingName: false,
        error: action.payload,
      };
    case CHANGE_ID:
      return {
        ...state,
        id: action.payload,
      };
    case FETCH_PROGRAM:
      return {
        ...state,
        isFetching: true,
      };
    case FETCH_PROGRAM_FULFILLED:
      return {
        ...state,
        isFetching: false,
        xmlCode: action.payload.content,
        id: action.payload.id,
        name: action.payload.name,
      };
    case FETCH_PROGRAM_REJECTED:
      return {
        ...state,
        isFetching: false,
        error: action.payload,
      };
    case SAVE_PROGRAM:
      return {
        ...state,
        isSaving: true,
      };
    case SAVE_PROGRAM_FULFILLED:
      return {
        ...state,
        isSaving: false,
        xmlCode: action.payload.content,
        id: action.payload.id,
        name: action.payload.name,
      };
    case SAVE_PROGRAM_REJECTED:
      return {
        ...state,
        isSaving: false,
        error: action.payload,
      };
    case CREATE_PROGRAM:
      return {
        ...state,
        isCreating: true,
      };
    case CREATE_PROGRAM_FULFILLED:
      return {
        ...state,
        isCreating: false,
        xmlCode: action.payload.content,
        id: action.payload.id,
        name: action.payload.name,
      };
    case CREATE_PROGRAM_REJECTED:
      return {
        ...state,
        isCreating: false,
        error: action.payload,
      };
    default:
      return state;
  }
}
