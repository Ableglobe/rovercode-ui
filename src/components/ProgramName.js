import React, { Component, Fragment } from 'react';
import { Confirm, Input } from 'semantic-ui-react';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies } from 'react-cookie';
import PropTypes from 'prop-types';

import { updateValidAuth } from '@/actions/auth';
import { changeName as actionChangeName } from '@/actions/code';

const mapStateToProps = ({ code }) => ({ code });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  changeName: (id, name) => {
    const changeNameAction = actionChangeName(id, name, {
      headers: {
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      },
    });
    return dispatch(changeNameAction).catch((error) => {
      if (error.response.status === 401) {
        // Authentication is no longer valid
        dispatch(updateValidAuth(false));
      }
    });
  },
});

class ProgramName extends Component {
  constructor(props) {
    super(props);

    this.state = {
      confirmOpen: false,
      editingName: null,
      previousPropName: null, // eslint-disable-line react/no-unused-state
    };
  }

  static getDerivedStateFromProps(props, state) {
    const { code } = props;
    const { previousPropName } = state;

    if (code.name !== previousPropName) {
      return {
        editingName: code.name,
        previousPropName: code.name,
      };
    }

    return null;
  }

  closeConfirm = () => this.setState({ confirmOpen: false })

  openConfirm = () => this.setState({ confirmOpen: true })

  handleChange = (e) => {
    this.setState({
      editingName: e.target.value,
    });
  }

  handleSave = () => {
    const { changeName, code } = this.props;
    const { editingName } = this.state;

    changeName(code.id, editingName);
    this.closeConfirm();
  }

  render() {
    const { code } = this.props;
    const { confirmOpen, editingName, previousPropName } = this.state;
    let actionProp = {};

    // Only show `Save` when the name has changed
    if (editingName !== previousPropName) {
      actionProp = {
        action: {
          content: 'Save',
          onClick: this.openConfirm,
        },
      };
    }

    return (
      <Fragment>
        <Input
          type="text"
          label="Name:"
          defaultValue={editingName}
          disabled={code.isReadOnly}
          onChange={this.handleChange}
          {...actionProp}
        />
        <Confirm
          content="Are you sure that you want to change the name of this program?"
          open={confirmOpen}
          onCancel={this.closeConfirm}
          onConfirm={this.handleSave}
        />
      </Fragment>
    );
  }
}

ProgramName.propTypes = {
  code: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    isReadOnly: PropTypes.bool,
  }).isRequired,
  changeName: PropTypes.func.isRequired,
};

export default hot(module)(withCookies(connect(mapStateToProps, mapDispatchToProps)(ProgramName)));
