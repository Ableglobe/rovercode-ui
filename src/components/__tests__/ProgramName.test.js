import React from 'react';
import { Input } from 'semantic-ui-react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Cookies } from 'react-cookie';

import { changeName } from '@/actions/code';
import ProgramName from '../ProgramName';

const cookiesValues = { auth_jwt: '1234' };
const cookies = new Cookies(cookiesValues);

describe('The Console component', () => {
  const mockStore = configureStore();
  const context = { cookies };
  let store;

  beforeEach(() => {
    store = mockStore({
      code: {
        name: 'test name',
      },
    });
    store.dispatch = jest.fn();
  });

  test('renders on the page with no errors', () => {
    const wrapper = mount(<ProgramName store={store} />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('displays name', () => {
    const wrapper = shallow(<ProgramName store={store} />, { context }).dive().dive();

    expect(wrapper.find(Input).length).toBe(1);
    expect(wrapper.find(Input).props().defaultValue).toBe('test name');
  });

  test('handles change', () => {
    const wrapper = shallow(<ProgramName store={store} />, { context }).dive().dive();

    wrapper.find(Input).simulate('change', { target: { value: 'new name' } });
    wrapper.update();

    expect(wrapper.find(Input).props().defaultValue).toBe('new name');
  });

  test('handles save', () => {
    const wrapper = shallow(<ProgramName store={store} />, { context }).dive().dive();

    wrapper.find(Input).props().action.onClick();
    wrapper.update();

    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(changeName('test name'));
  });
});
