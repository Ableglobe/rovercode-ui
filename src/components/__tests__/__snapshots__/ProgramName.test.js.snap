// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Console component renders on the page with no errors 1`] = `
<withCookies(Component)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Console)
    allCookies={Object {}}
    cookies={
      Cookies {
        "HAS_DOCUMENT_COOKIE": true,
        "changeListeners": Array [
          [Function],
        ],
        "cookies": Object {},
        "hooks": undefined,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <Console
      allCookies={Object {}}
      changeName={[Function]}
      code={
        Object {
          "name": "test name",
        }
      }
      cookies={
        Cookies {
          "HAS_DOCUMENT_COOKIE": true,
          "changeListeners": Array [
            [Function],
          ],
          "cookies": Object {},
          "hooks": undefined,
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
      storeSubscription={
        Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [MockFunction],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        }
      }
    >
      <Input
        defaultValue="test name"
        label="Name:"
        onChange={[Function]}
        type="text"
      >
        <div
          className="ui labeled input"
        >
          <Label
            className="label"
            content="Name:"
          >
            <div
              className="ui label label"
              onClick={[Function]}
            >
              Name:
            </div>
          </Label>
          <input
            defaultValue="test name"
            onChange={[Function]}
            type="text"
          />
        </div>
      </Input>
      <Confirm
        cancelButton="Cancel"
        confirmButton="OK"
        content="Are you sure that you want to change the name of this program?"
        onCancel={[Function]}
        onConfirm={[Function]}
        open={false}
        size="small"
      >
        <Modal
          centered={true}
          closeOnDimmerClick={true}
          closeOnDocumentClick={false}
          dimmer={true}
          eventPool="Modal"
          onClose={[Function]}
          open={false}
          size="small"
        >
          <Portal
            closeOnDocumentClick={false}
            closeOnEscape={true}
            eventPool="Modal"
            mountNode={<body />}
            onClose={[Function]}
            onMount={[Function]}
            onOpen={[Function]}
            onUnmount={[Function]}
            open={false}
            openOnTriggerClick={true}
          />
        </Modal>
      </Confirm>
    </Console>
  </Connect(Console)>
</withCookies(Component)>
`;
